docs/operator-manual/applicationset/Progressive-Rollouts.md

```diff
@@ -0,0 +1,95 @@
+
+## Strategies
+
+* AllAtOnce (default)
+* RollingUpdate
+
+### AllAtOnce
+This default Application update behavior is unchanged from the original ApplicationSet implementation.
+
+All Applications managed by the ApplicationSet resource are updated simultaneously when the ApplicationSet is updated.
+
+### RollingUpdate
+This update strategy allows you to group Applications by labels present on the generated Application resources.
+When the ApplicationSet changes, the changes will be applied to each group of Application resources sequentially.
+
+* Application groups are selected by `matchExpressions`.
+* All `matchExpressions` must be true for an Application to be selected (AND behavior).
+* The `In` operator must match at least one value to be considered true (OR behavior).
+* All Applications in each group must become Healthy before the ApplicationSet controller will proceed to update the next group of Applications.
+* The number of simultaneous Application updates in a group will not exceed its `maxUpdate` parameter (0 or undefined means unbounded).
+
+#### Example
+The following example illustrates how to stage a progressive rollout over Applications with explicitly configured environment labels.
+```
+---
+apiVersion: argoproj.io/v1alpha1
+kind: ApplicationSet
+metadata:
+  name: guestbook
+spec:
+  generators:
+  - list:
+      elements:
+      - cluster: engineering-dev
+        url: https://1.2.3.4
+        env: dev
+      - cluster: engineering-qa
+        url: https://2.4.6.8
+        env: qa
+      - cluster: engineering-prod
+        url: https://9.8.7.6/
+        env: prod
+  strategy:
+    type: RollingUpdate
+    rollingUpdate:
+      steps:
+        - matchExpressions:
+            - key: env
+              operator: In
+              values:
+                - dev
+          maxUpdate: 0 # if undefined or 0, all applications matched are updated together
+        - matchExpressions:
+            - key: env
+              operator: In
+              values:
+                - qa
+        - matchExpressions:
+            - key: env
+              operator: In
+              values:
+                - prod
+          maxUpdate: 1 # maxUpdate supports both integer and percentage string values
+  template:
+    metadata:
+      name: '{{cluster}}-guestbook'
+      labels:
+        env: '{{env}}'
+    spec:
+      source:
+        repoURL: https://github.com/infra-team/cluster-deployments.git
+        targetRevision: HEAD
+        path: guestbook/{{cluster}}
+      destination:
+        server: '{{url}}'
+        namespace: guestbook
+```
+
+#### Limitations
+The RollingUpdate strategy does not enforce sync order for external changes. Basically, if the ApplicationSet spec does not change, the RollingUpdate strategy has no mechanism available to control the sync order of the Applications it manages. The managed Applications will sync on their own (if autosync is enabled), and your desired rollingUpdate spec will be ignored.
```

Comment:
Is this true even if a parameter value changes (e.g. from a git file generator update), i.e. changing the evaluated template values?

Reply:
Thanks for catching this edge case. It's not handled properly right now, but I should be able to handle it by tracking a hash of the generated Application resource to know if an individual Application has changed even though the ApplicationSet spec has not.

Reply:
Application hash tracking and an experimental feature gate are both now implemented. Changes to the generated Application (but not the ApplicationSet, such as templated cluster secret labels) are now caught and handled consistently.

This commit includes the diff for the original implementation and this new behavior.
https://github.com/argoproj/argo-cd/pull/10048/commits/dca8a49ca7bc37a0a8ee817877a6f3f6a3218cb0